jmix-flowui/flowui/src/main/java/io/jmix/flowui/data/binding/impl/HtmlContainerReadonlyDataBindingImpl.java

```diff
@@ -0,0 +1,179 @@
+            Object propertyValue = itemChangeEvent.getItem() != null
+                    ? EntitySystemAccess.getEntityEntry(itemChangeEvent.getItem()).getAttributeValue(property)
+                    : null;
+            updateHtmlContainerText(htmlContainer, formatter, propertyValue);
+        });
+        DataBindingListenerCleaner itemChangeCleaner = itemChangeSubscription::remove;
+        listenersRegistrations.add(new Pair<>(htmlContainer, itemChangeCleaner));
+    }
+
+    @Override
+    public void bind(HtmlContainer htmlComponent, CollectionContainer<?> dataContainer) {
+        bind(htmlComponent, dataContainer, new CollectionFormatter(metadataTools));
+    }
+
+    @Override
+    public void bind(HtmlContainer htmlContainer, CollectionContainer<?> dataContainer,
+                     Formatter<Collection<?>> formatter) {
+        checkNotNullArgument(htmlContainer);
+        checkNotNullArgument(dataContainer);
+        checkNotNullArgument(formatter);
+
+        unbind(htmlContainer, true);
+
+        updateHtmlContainerText(htmlContainer, formatter, dataContainer.getItems());
+
+        Subscription subscription = dataContainer.addCollectionChangeListener(collectionChangeEvent ->
+                updateHtmlContainerText(htmlContainer, formatter, dataContainer.getItems()));
+
+        DataBindingListenerCleaner dataBindingListenerCleaner = subscription::remove;
+        listenersRegistrations.add(new Pair<>(htmlContainer, dataBindingListenerCleaner));
+    }
+
+    public void unbind(HtmlContainer htmlContainer) {
+        unbind(htmlContainer, false);
+    }
+
+    protected void updateHtmlContainerText(HtmlContainer htmlContainer,
+                                           Formatter<Object> formatter,
+                                           @Nullable Object value) {
+        htmlContainer.setText(formatter.apply(value));
+    }
+
+    protected void updateHtmlContainerText(HtmlContainer htmlContainer,
+                                           Formatter<Collection<?>> formatter,
+                                           @Nullable Collection<?> value) {
+        htmlContainer.setText(formatter.apply(value));
+    }
+
+    protected void unbind(HtmlContainer htmlContainer, boolean warnIfBound) {
+        checkNotNullArgument(htmlContainer);
+
+        Iterator<Pair<HtmlContainer, DataBindingListenerCleaner>> iterator = listenersRegistrations.iterator();
+        while (iterator.hasNext()) {
+            Pair<HtmlContainer, DataBindingListenerCleaner> htmlContainerAndCleaner = iterator.next();
+            if (htmlContainer.equals(htmlContainerAndCleaner.getFirst())) {
+                htmlContainerAndCleaner.getSecond().remove();
+                iterator.remove();
+
+                if (warnIfBound) {
+                    log.warn("Silent unbind. HtmlContainer with id = '{}' was already bounden with data source.",
+                            htmlContainer.getId());
+                }
+            }
+        }
```

Comment:
When `list` will be replaced to `map` then you can simply use the following construction here:

```suggestion
        Runnable remove = listenersRegistrations.get(htmlContainer);
        if (remove == null) {
            return;
        }

        remove.run();
        listenersRegistrations.remove(htmlContainer);
        
        if (warnIfBound) {
            log.warn("Silent unbind. HtmlContainer with id = '{}' was already bounden with data source.",
                    htmlContainer.getId());
        }
```